apps/from_usm_ndarray/usm_ndarray_dlpack.cpp

```diff
@@ -0,0 +1,99 @@
+#include <CL/sycl.hpp>
+#include <cstdint>
+#include <iostream>
+#include "dlpack/dlpack.h"
+
+extern "C" {
+void Give(DLManagedTensor dl_managed_tensor);
+void Finalize();
+void FreeHandle();
+}
+
+namespace {
+
+DLManagedTensor *given = nullptr;
+
+void display(DLManagedTensor mt) {
+    DLTensor tensor = mt.dl_tensor;
+    DLDevice dldev = tensor.device;
+    DLDataType dtype = tensor.dtype;
+    int i, ndim = tensor.ndim;
+    std::cout << "On C++ side" << std::endl;
+    std::cout << "Data: " << std::hex << tensor.data << std::dec << std::endl;
+    std::cout << "Device (" << dldev.device_type << ", " << dldev.device_id << ")" << std::endl;
+    std::cout << "Type description: { 'code': " << static_cast<int>(dtype.code)
+	      << ", 'bits': " << static_cast<int>(dtype.bits)
+	      << ", 'lanes': " << static_cast<int>(dtype.lanes) << " }"  << std::endl;
+    std::cout << "ndim: " << ndim << std::endl;
+    std::cout << "shape: (";
+    {
+	int64_t* sh = tensor.shape;
+	for(int i=0; i < ndim; ++i) {
+	    std::cout << sh[i];
+	    std::cout << ((i+1) < ndim ? ", " : "");
+	}
+    }
+    std::cout << ")" << std::endl;
+
+    std::cout << "strides: (";
+    {
+	int64_t* st = tensor.strides;
+	for(int i=0; i < ndim; ++i) {
+	    std::cout << st[i];
+	    std::cout << ((i+1) < ndim ? ", " : "");
+	}
+    }
+    std::cout << ")" << std::endl;
+
+    sycl::device *d_ptr = nullptr;
+
+    switch(dldev.device_type) {
+    case kDLONEAPI_GPU:
+	d_ptr = new sycl::device(sycl::ONEAPI::filter_selector("gpu:" + std::to_string(dldev.device_id)));
+	break;
+    case kDLONEAPI_CPU:
+	d_ptr = new sycl::device(sycl::ONEAPI::filter_selector("cpu:" + std::to_string(dldev.device_id)));
+	break;
+    case kDLONEAPI_ACCELERATOR:
+	d_ptr = new sycl::device(sycl::ONEAPI::filter_selector("accelerator:" + std::to_string(dldev.device_id)));
+	break;
+    default:
+	throw std::runtime_error("Can not handler these types");
+    }
+
+    sycl::queue q(*d_ptr);   // queue uses default context per root device
+    sycl::context ctxt = q.get_context();
+    delete d_ptr;
+
+    sycl::usm::alloc kind = sycl::get_pointer_type(tensor.data, ctxt);
```

Comment:
In a zero-copy exchange of a USM allocation between two applications, is there additional handling that generated code needs to employ when the allocation kind is `shared` vs `device`? Said differently, if a backend consumes a dltensor with  kDLONEAPI_GPU device type, will it be sufficient for the runtime to consume allocations of these two kinds interchangeably?

Reply:
Once the importer figures out the SYCL context associated with USM allocation, any SYCL kernel submitted to a queue associated with such a context can access USM memory irrespective of the USM allocation type.  The runtime takes care of that.

Once USM allocation is made, there is no way to change its allocation type. The type of USM allocation is queried from the pointer and the sycl context it is associated with.  If query is done against a different SYCL context, the returned type is "unknown"

In this particular application, line 68 serves to demonstrate that query worked correctly, and thus correct context was reconstructed.